dali/kernels/imgproc/warp/block_warp.cuh

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_WARP_BLOCK_WARP_CUH_
+#define DALI_KERNELS_IMGPROC_WARP_BLOCK_WARP_CUH_
+
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/imgproc/sampler.h"
+#include "dali/kernels/imgproc/warp/warp_setup.cuh"
+#include "dali/kernels/imgproc/warp/mapping_traits.h"
+
+namespace dali {
+namespace kernels {
+namespace warp {
+
+template <typename Mapping, std::size_t dim>
+DALI_HOST_DEV
+enable_if_t<is_fp_mapping<Mapping>::value, vec<dim>> map_coords(const Mapping &m, ivec<dim> pos) {
+  return m(pos + 0.5f);
+}
+
+template <typename Mapping, std::size_t dim>
+DALI_HOST_DEV
+enable_if_t<!is_fp_mapping<Mapping>::value, ivec<dim>> map_coords(const Mapping &m, ivec<dim> pos) {
+  return m(pos);
+}
+
+template <DALIInterpType interp_type, typename Mapping,
+          int ndim, typename OutputType, typename InputType,
+          typename BorderValue>
+__device__ void BlockWarp(
```

Comment:
I wonder if it should be "2D" somewhere in the name. As in the future we may need to have 3D warp.

Reply:
Not really. I'll just add an overload with `SampleDesc<3, >`, `BlockDesc<3>` and C++ overload resolution will do the trick.